### PR TITLE
Enable nullable annotations

### PR DESCRIPTION
**Pull request by DaleStan**
This enables nullable annotations for all four non-test projects and removes the corresponding warnings. I found and fixed a few bugs in the process, and did some refactoring when I happened to notice cases where static-to-instance changes could remove some of the null-forgiving operators. There's probably more refactoring to do; I know Dorus wants to do more work with `FlatHierarchy`, for example.

Unfortunately, this does cause conflicts with most of the other PRs. We could take only some of the nullable changes, but I thought having nullable in some projects but not others would lead to confusion.

### CODE REVIEW

**Comment by have-fun-was-taken** [comment]
Thank you for the PR! 

Could you please elaborate on the reason behind these changes? You can just link a couple of inspections that you fixed so we can read on what was wrong, or you can explain it in your own words.

I noticed that some null checks became unnecessary and were removed, but I wanted to know if there was anything else that motivated this change.

File: Yafc/Windows/ProjectPageSettingsPanel.cs [hunk 1]
```diff
@@ -151,9 +153,9 @@ private class ExportRecipe {
 
             public ExportRecipe(RecipeRow row) {
                 Recipe = row.recipe.name;
-                Building = row.entity.name;
+                Building = row.entity?.name ?? "<No building selected>";
```
Comment (DaleStan) [inline Yafc/Windows/ProjectPageSettingsPanel.cs:156]: Two of the bugs I remember fixing are this line and `row.fuel?.name`, two lines down. It is possible to arrange things so a recipe has no available crafting entities, which means that both row.entity and row.fuel will be null. Most of the code checks those for null, but these two lines didn't.

File: Yafc/Windows/PreferencesScreen.cs [hunk 2]
```diff
@@ -94,15 +94,27 @@ public class PreferencesScreen : PseudoScreen {
...
+                    gui.BuildObjectSelectDropDown(list, DataUtils.DefaultOrdering, selectItem, text, width: width);
+                }
+            }
+        }
+
+        /// <summary>Add a GUI element that opens a popup to allow the user to choose from the <paramref name="list"/>, which triggers <paramref name="selectItem"/>.
+        /// An additional "clear" or "none" option will also be displayed.</summary>
+        /// <param name="text">Label to show.</param>
+        /// <param name="width">Width of the popup. Make sure it is wide enough to fit text!</param>
+        private static void ChooseObjectWithNone<T>(ImGui gui, string text, T[] list, T? current, Action<T?> selectItem, float width = 20f) where T : FactorioObject {
```
Comment (DaleStan) [inline Yafc/Windows/PreferencesScreen.cs:113]: Here, I separated `ChooseObject` and `ChooseObjectWithNone` (along with creating several other `...WithNone` methods) so the compiler could check the `selectItem` argument. If the UI is displaying a 'none' or 'clear' item, `selectItem` must accept a null value. Nullable annotations don't have the ability to tell the compiler about that with an allowNone parameter, so I moved that into the method name instead.

**Comment by DaleStan** [comment]
In addition to the previous comments, one of the biggest advantages, in my opinion, is that all methods now specify whether they can accept or return null values, in a way the compiler can understand. This way the compiler can make sure we're inserting appropriate null checks, reducing the chance of bugs.

**Review by have-fun-was-taken — APPROVED** [review]

**Review by have-fun-was-taken — CHANGES_REQUESTED** [review]
Loading an existing PyAE project fails with the following error:

grafik

**Comment by have-fun-was-taken** [comment]
Here's the Factorio save file to sync mods and the project which loading triggered the error.
[dorusNullableSave.zip](https://github.com/user-attachments/files/15525038/dorusNullableSave.zip)
[project.zip](https://github.com/user-attachments/files/15525040/project.zip)

**Review by have-fun-was-taken — APPROVED** [review]
Thank you for the fix of the bug! We can wait for someone else to take a look at the code, or we can merge it now. The choice is yours.

**Comment by DaleStan** [comment]
That happened because I added `name` as a constructor parameter, and YAFC could read
```json
  "sharedModuleTemplates": [
    {
      "name": "Nothing", // Any number of consecutive constructor parameters is fine,
      "template": {
        "beacon": null,
        "list": [],
        "beaconList": []
      },
      "icon": null,
      "filterEntities": []
    }
  ],
```
but not
```json
  "sharedModuleTemplates": [
    {
      "template": {
        "beacon": null,
        "list": [],
        "beaconList": []
      },
      "icon": null,
      "name": "Nothing", // but constructor parameters couldn't be found if they appeared after non-parameters.
      "filterEntities": []
    }
  ],
```

It can now find constructor parameters even if they appear after property values.

**Pull request merged**